kafka-impl/src/main/java/io/streamnative/pulsar/handlers/kop/storage/PartitionLog.java

```diff
@@ -0,0 +1,257 @@
+        KafkaTopicManager topicManager = appendRecordsContext.getTopicManager();
+        final long beforeRecordsProcess = time.nanoseconds();
+        try {
+            MemoryRecords validRecords = validateRecords(version, fullPartitionName, records);
+            validRecords.batches().forEach(batch->{
+                if (batch.sizeInBytes() > kafkaConfig.getMaxMessageSize()) {
+                    throw new RecordTooLargeException(String.format("Message batch size is %s "
+                                    + "in append to partition %s which exceeds the maximum configured size of %s .",
+                            batch.sizeInBytes(), topicPartition, kafkaConfig.getMaxMessageSize()));
+                }
+            });
+            // Append Message into pulsar
+            final CompletableFuture<Optional<PersistentTopic>> topicFuture =
+                    topicManager.getTopic(fullPartitionName);
+            if (topicFuture.isCompletedExceptionally()) {
+                topicFuture.exceptionally(e -> {
+                    appendFuture.completeExceptionally(e);
+                    return Optional.empty();
+                });
+                return appendFuture;
+            }
+            if (topicFuture.isDone() && !topicFuture.getNow(Optional.empty()).isPresent()) {
+                appendFuture.completeExceptionally(Errors.NOT_LEADER_FOR_PARTITION.exception());
+                return appendFuture;
+            }
+            final Consumer<Optional<PersistentTopic>> persistentTopicConsumer = persistentTopicOpt -> {
+                if (!persistentTopicOpt.isPresent()) {
+                    appendFuture.completeExceptionally(Errors.NOT_LEADER_FOR_PARTITION.exception());
+                    return;
+                }
+                // TODO: validateMessagesAndAssignOffsets here.
+
+                final EncodeRequest encodeRequest = EncodeRequest.get(validRecords);
+                if (entryFormatter instanceof KafkaMixedEntryFormatter) {
+                    final ManagedLedger managedLedger = persistentTopicOpt.get().getManagedLedger();
+                    final long logEndOffset = MessageMetadataUtils.getLogEndOffset(managedLedger);
+                    encodeRequest.setBaseOffset(logEndOffset);
+                }
+
+                final EncodeResult encodeResult = entryFormatter.encode(encodeRequest);
+                encodeRequest.recycle();
+                requestStats.getProduceEncodeStats().registerSuccessfulEvent(
+                        time.nanoseconds() - beforeRecordsProcess, TimeUnit.NANOSECONDS);
+                appendRecordsContext.getStartSendOperationForThrottling()
+                        .accept(encodeResult.getEncodedByteBuf().readableBytes());
+                if (log.isDebugEnabled()) {
+                    log.debug("Produce messages for topic {} partition {}",
+                            topicPartition.topic(), topicPartition.partition());
+                }
```

Comment:
Actually I think this debug logs doesn't help. The original debug logs are

```java
                if (log.isDebugEnabled()) {
                    log.debug("[{}] Request {}: Produce messages for topic {} partition {}, "
                                    + "request size: {} ", ctx.channel(), produceHar.getHeader(),
                            topicPartition.topic(), topicPartition.partition(), numPartitions);
                }
```

BTW, the `numPartitions` is wrong, it should not represent the request size.

I think you can just remove these lines or pass the channel context and request header to `AppendRecordsContext`.